### PR TITLE
feat: bootstrap controller charm on CAAS

### PR DESCRIPTION
**Pull request by SimonRichardson**
The IAAS bootstrap path has long supported deploying a local controller charm (i.e. a .charm file from the local filesystem rather than pulling from Charmhub) and correctly forwards constraints to the application service. The CAAS path was missing both of these - it explicitly rejected local charm paths with a "not supported" error, and was silently dropping constraints. This asymmetry made CAAS a second-class citizen for controller development and debugging workflows.

## QA steps

```sh
$ make microk8s-operator-update
$ juju bootstrap microk8s ops --controller-charm-path=./juju-controller_ubuntu@24.04-amd64.charm
```

## Links

**Jira card:** [JUJU-10075](https://warthogs.atlassian.net/browse/JUJU-10075)

[JUJU-10075]: https://warthogs.atlassian.net/browse/JUJU-10075?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by jack-w-shaw — APPROVED** [review]

File: internal/provider/kubernetes/bootstrap.go
Comment (nvinuesa) [inline internal/provider/kubernetes/bootstrap.go:1036]: This is of course not you, but this method should have been named `GetControllerPodName()`.

File: internal/provider/kubernetes/bootstrap_test.go [hunk 1]
```diff
@@ -292,6 +293,32 @@ func (s *bootstrapSuite) TestGetControllerSvcSpec(c *tc.C) {
...
+			break
+		}
+	}
+	c.Assert(apiServer, tc.NotNil)
+	c.Assert(apiServer.Args, tc.HasLen, 2)
+
+	startup := apiServer.Args[1]
+	c.Check(startup, tc.Contains, "mkdir -p $JUJU_DATA_DIR/charms")
+	c.Check(startup, tc.Contains, "until test -e $JUJU_DATA_DIR/charms/controller.charm; do sleep 1; done")
+	c.Check(startup, tc.Contains, "$JUJU_TOOLS_DIR/jujud bootstrap-state --data-dir $JUJU_DATA_DIR --debug --timeout 10m0s")
```
Comment (nvinuesa) [inline internal/provider/kubernetes/bootstrap_test.go]: You'll need to update this after rebasing.

File: internal/provider/kubernetes/bootstrap.go [hunk 2]
```diff
@@ -325,6 +328,88 @@ func newControllerStack(
...
+		return errors.Annotate(err, "installing local controller charm")
+	}
+	return nil
+}
+
+func (c *controllerStack) uploadLocalControllerCharmWithRetry(ctx context.Context, podName string) error {
+	if !isLocalControllerCharmPath(c.pcfg.Bootstrap.ControllerCharmPath) {
+		return nil
+	}
+	return retry.Call(retry.CallArgs{
```
Comment (gfouillet) [inline internal/provider/kubernetes/bootstrap.go:404]: **todo:** The `retry.CallArgs` is missing both `IsFatalError` and `Stop`.

Without `IsFatalError`, permanent errors (e.g. `controllerExecClient` returning `errors.NotValidf("empty kubernetes rest config")`) are retried 12 times, wasting ~12 seconds. Without `Stop: ctx.Done()`, the retry loop doesn't honour context cancellation between attempts — up to ~11 seconds of unnecessary blocking after cancellation.

```go
return retry.Call(retry.CallArgs{
    Attempts: 12,
    Delay:    time.Second,
    Clock:    c.broker.clock,
    Func: func() error {
        return c.uploadLocalControllerCharm(ctx, podName)
    },
})
```

Also, the magic numbers `12` and `time.Second` are unexplained — consider named constants or a comment. Consider adding `IsFatalError` to classify permanent errors, `Stop: ctx.Done()` for cancellation, and `NotifyFunc` for retry logging.

File: internal/provider/kubernetes/bootstrap.go [hunk 3]
```diff
@@ -325,6 +328,88 @@ func newControllerStack(
...
+func isLocalControllerCharmPath(charmPath string) bool {
+	return strings.HasPrefix(charmPath, "/") || strings.HasPrefix(charmPath, "./") ||
+		strings.HasPrefix(charmPath, "../")
+}
+
+func (c *controllerStack) localControllerCharmArchivePath() string {
+	return c.pathJoin(c.pcfg.DataDir, "charms", environsbootstrap.ControllerCharmArchive)
+}
+
+func (c *controllerStack) uploadLocalControllerCharm(ctx context.Context, podName string) error {
```
Comment (gfouillet) [inline internal/provider/kubernetes/bootstrap.go:345]: **todo:** No test exists for `uploadLocalControllerCharm` — the core function that copies a charm to a pod via k8s exec/copy (mkdir → copy → chmod → mv) has zero coverage, including all failure paths (exec client error, mkdir error, copy error, chmod error, mv error) and the non-local early-return.

The function is difficult to test without refactoring `controllerExecClient` to be injectable. Consider adding an `execClientFactory` func field on `controllerStack` (overridable in tests via `export_test.go`).

File: internal/provider/kubernetes/bootstrap.go [hunk 4]
```diff
@@ -325,6 +328,88 @@ func newControllerStack(
...
+		return errors.Annotate(err, "setting local controller charm permissions")
+	}
+
+	if err := execCommand([]string{"mv", "-f", uploadPath, archivePath}); err != nil {
+		return errors.Annotate(err, "installing local controller charm")
+	}
+	return nil
+}
+
+func (c *controllerStack) uploadLocalControllerCharmWithRetry(ctx context.Context, podName string) error {
```
Comment (gfouillet) [inline internal/provider/kubernetes/bootstrap.go:400]: **todo:** No test exists for `uploadLocalControllerCharmWithRetry` — the retry wrapper (12 attempts, 1s delay) is untested.

With an injectable exec client factory, test: (1) first 2 calls fail, 3rd succeeds — assert nil return and exactly 3 calls; (2) all 12 attempts fail — assert non-nil error; (3) non-local charm path — assert nil return with zero calls.

File: internal/provider/kubernetes/bootstrap_test.go [hunk 5]
```diff
@@ -292,6 +293,32 @@ func (s *bootstrapSuite) TestGetControllerSvcSpec(c *tc.C) {
 	}
```
Comment (gfouillet) [inline internal/provider/kubernetes/bootstrap_test.go:301]: **todo:** `TestBootstrap` (line 322) doesn't set `ControllerCharmPath`, so the upload function returns nil early — the local charm upload integration path through `createControllerStatefulset` is completely untested.

While `TestControllerSpecWaitsForLocalControllerCharm` covers the container spec, it doesn't exercise the actual upload in `createControllerStatefulset`. A variant of `TestBootstrap` should set `s.pcfg.Bootstrap.ControllerCharmPath = "/tmp/controller.charm"` and verify that `uploadLocalControllerCharmWithRetry` is called and that an upload failure returns an error annotated with "uploading local controller charm".

File: internal/bootstrap/deployer_caas_test.go [hunk 6]
```diff
@@ -49,6 +57,80 @@ func (s *deployerCAASSuite) TestControllerCharmBase(c *tc.C) {
...
+				"controller-url":        "wss://obscura.com:1234/api",
+			},
+			ApplicationSettings: domainapplication.ApplicationSettings{
+				Trust: true,
+			},
+			ApplicationStatus: &status.StatusInfo{
+				Status: status.Unset,
+				Since:  new(now),
+			},
+			Constraints:  constraints.Value{},
```
Comment (gfouillet) [inline internal/bootstrap/deployer_caas_test.go]: **todo:** `TestAddCAASControllerApplication` expects `Constraints: constraints.Value{}` (empty) because `newConfig` sets `Constraints: constraints.Value{}` at `package_test.go:79`. The PR's key change (`Constraints: b.constraints`) is not meaningfully tested — the test would pass even if the deployer always sent empty constraints.

Set `cfg.Constraints = constraints.Value{Arch: new("arm64"), Mem: new(4096)}` in the test config and update the expected `AddApplicationArgs` to match, verifying the deployer forwards the constraints it was configured with.

File: internal/provider/kubernetes/bootstrap.go [hunk 7]
```diff
@@ -325,6 +328,88 @@ func newControllerStack(
 	return cs, nil
 }
 
+func isLocalControllerCharmPath(charmPath string) bool {
```
Comment (gfouillet) [inline internal/provider/kubernetes/bootstrap.go:335]: **nitpick:** `isLocalControllerCharmPath` is a verbatim duplicate of `refresher.IsLocalURL` (`cmd/juju/application/refresher/refresher.go:213`) without a cross-reference comment.

The duplication is architecturally justified — `internal/provider/kubernetes` cannot import `cmd/`. However, a comment should be added: `// mirrors refresher.IsLocalURL (cmd/juju/application/refresher/refresher.go)`. Also, no test covers the three prefix cases (`/`, `./`, `../`) or edge cases (empty string, non-local paths). Consider exporting via `export_test.go` for a table-driven test.

File: internal/provider/kubernetes/bootstrap.go [hunk 8]
```diff
@@ -325,6 +328,88 @@ func newControllerStack(
...
+
+	execClient, err := c.controllerExecClient()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	archivePath := c.localControllerCharmArchivePath()
+	uploadPath := archivePath + ".uploading"
+	archiveDir := path.Dir(archivePath)
+	execCommand := func(commands []string) error {
```
Comment (gfouillet) [inline internal/provider/kubernetes/bootstrap.go:359]: **suggestion:** The `execCommand` closure allocates `stdout` and `stderr` buffers but never includes them in error diagnostics. When `mkdir`/`chmod`/`mv` fails, the pod's stderr output (which typically contains the useful diagnostic) is silently discarded.

Consider including `stderr.String()` in the annotated error or logging it:
```go
return errors.Annotatef(err, "creating local controller charm directory: %s", stderr.String())
```

... Or log it.

File: internal/provider/kubernetes/bootstrap.go [hunk 9]
```diff
@@ -325,6 +328,88 @@ func newControllerStack(
...
+	}
+
+	execClient, err := c.controllerExecClient()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	archivePath := c.localControllerCharmArchivePath()
+	uploadPath := archivePath + ".uploading"
+	archiveDir := path.Dir(archivePath)
```
Comment (gfouillet) [inline internal/provider/kubernetes/bootstrap.go:358]: **note:** `path.Dir(archivePath)` uses stdlib `path.Dir` which cleans paths per `path.Clean` semantics, while the custom `pathJoin` method (line 425) does not clean paths. If we do not replace `pathJoin` by `path.Join` it will causes an inconsistency.

File: internal/provider/kubernetes/bootstrap_test.go [hunk 10]
```diff
@@ -292,6 +293,32 @@ func (s *bootstrapSuite) TestGetControllerSvcSpec(c *tc.C) {
...
+		}
+	}
+	c.Assert(apiServer, tc.NotNil)
+	c.Assert(apiServer.Args, tc.HasLen, 2)
+
+	startup := apiServer.Args[1]
+	c.Check(startup, tc.Contains, "mkdir -p $JUJU_DATA_DIR/charms")
+	c.Check(startup, tc.Contains, "until test -e $JUJU_DATA_DIR/charms/controller.charm; do sleep 1; done")
+	c.Check(startup, tc.Contains, "$JUJU_TOOLS_DIR/jujud bootstrap-state --data-dir $JUJU_DATA_DIR --debug --timeout 10m0s")
+	c.Check(strings.Contains(startup, "test -e $JUJU_DATA_DIR/agents/controller-0/agent.conf ||"), tc.IsFalse)
```
Comment (gfouillet) [inline internal/provider/kubernetes/bootstrap_test.go]: **nitpick:** `c.Check(strings.Contains(startup, "..."), tc.IsFalse)` would be more idiomatic as `c.Check(startup, tc.Not(tc.Contains), "...")` — it produces a more descriptive failure message showing the actual string content. The preceding assertions already use `tc.Contains` for positive checks.

File: internal/provider/kubernetes/bootstrap.go [hunk 11]
```diff
@@ -325,6 +328,88 @@ func newControllerStack(
...
+		Attempts: 12,
+		Delay:    time.Second,
+		Clock:    c.broker.clock,
+		Func: func() error {
+			return c.uploadLocalControllerCharm(ctx, podName)
+		},
+	})
+}
+
+func (c *controllerStack) controllerExecClient() (k8sexec.Executor, error) {
```
Comment (gfouillet) [inline internal/provider/kubernetes/bootstrap.go:421]: **todo:** No test for `controllerExecClient` — the nil restConfig guard (returns `errors.NotValidf("empty kubernetes rest config")`) is untested. Export the function in `export_test.go` and test the nil case.

File: internal/provider/kubernetes/bootstrap_test.go [hunk 12]
```diff
@@ -292,6 +293,32 @@ func (s *bootstrapSuite) TestGetControllerSvcSpec(c *tc.C) {
 	}
 }
 
+func (s *bootstrapSuite) TestControllerSpecWaitsForLocalControllerCharm(c *tc.C) {
```
Comment (gfouillet) [inline internal/provider/kubernetes/bootstrap_test.go:304]: **suggestion:** `TestControllerSpecWaitsForLocalControllerCharm` only tests the local charm path. Consider adding a companion test that leaves `ControllerCharmPath` empty (or sets it to `ch:juju-controller`) and asserts the poll-wait loop (`until test -e ...`) is absent — explicitly guarding against the wait logic leaking into the non-local path.

File: cmd/juju/commands/bootstrap.go [hunk 13]
```diff
@@ -664,9 +664,6 @@ func (c *bootstrapCommand) Run(ctx *cmd.Context) (resultErr error) {
 	}
 
 	isCAASController = jujucloud.CloudIsCAAS(cloud)
```
Comment (gfouillet) [inline cmd/juju/commands/bootstrap.go:666]: **todo:** The removed CAAS local charm rejection has no replacement regression test confirming `--controller-charm-path` is now accepted on CAAS clouds. The existing `TestBootstrapWithLocalControllerCharm` uses "dummy" cloud which is IAAS type. Consider adding a test with a k8s-type cloud to guard against the rejection being reintroduced.

**Review by nvinuesa — APPROVED** [review]
Uploading the charm from inside the provider felt a bit counterintuitive when I read it, but it is true that on machines we do the same (except cloud-init holds it) and the bootstrap worker seeds it.
LGTM

File: internal/provider/kubernetes/bootstrap.go
Comment (gfouillet) [inline internal/provider/kubernetes/bootstrap.go]: **todo:** remove me. Use `path.Join`

Comment (SimonRichardson) [reply internal/provider/kubernetes/bootstrap.go:1036]: Agreed, we should probably do a sweep of this at some point.
